src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/ModuleLoader.cs

```diff
@@ -0,0 +1,102 @@
+// or implied. See the License for the specific language governing permissions
+// and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Reactive.Subjects;
+
+namespace MorganStanley.ComposeUI.ModuleLoader;
+
+internal sealed class ModuleLoader : IModuleLoader, IAsyncDisposable
+{
+    private readonly Subject<LifetimeEvent> _lifetimeEvents = new();
+    private readonly ConcurrentDictionary<Guid, IModuleInstance> _modules = new();
+    private readonly Dictionary<string, IModuleRunner> _moduleRunners;
+    private readonly IModuleCatalog _moduleCatalog;
+    private readonly IReadOnlyList<IStartupAction> _startupActions;
+
+    public ModuleLoader(
+        IModuleCatalog moduleCatalog,
+        IEnumerable<IModuleRunner> moduleRunners,
+        IEnumerable<IStartupAction> startupActions)
+    {
+        ArgumentNullException.ThrowIfNull(moduleCatalog);
+        ArgumentNullException.ThrowIfNull(moduleRunners);
+        ArgumentNullException.ThrowIfNull(startupActions);
+
+        _moduleCatalog = moduleCatalog;
+        _moduleRunners = moduleRunners.GroupBy(runner => runner.ModuleType).ToDictionary(g => g.Key, g => g.First());
+        _startupActions = new List<IStartupAction>(startupActions);
+    }
+
+    public IObservable<LifetimeEvent> LifetimeEvents => _lifetimeEvents;
+
+    public async Task<IModuleInstance> StartModule(StartRequest request)
+    {
+        var manifest = _moduleCatalog.GetManifest(request.ModuleId);
+        if (manifest == null)
+        {
+            throw new Exception($"Unknown Module id: {request.ModuleId}");
+        }
+
+        if (!_moduleRunners.TryGetValue(request.ModuleId, out var moduleRunner))
+        {
+            throw new Exception($"No module runner available for {manifest.ModuleType} module type");
+        }
+
+        Guid instanceId = Guid.NewGuid();
+        var moduleInstance = new ModuleInstance(instanceId, manifest, request);
+        _modules.TryAdd(instanceId, moduleInstance);
+
+        _lifetimeEvents.OnNext(new LifetimeEvent.Starting(moduleInstance));
+        var startupContext = new StartupContext(request);
+
+        var pipeline = _startupActions
+            .Reverse()
+            .Aggregate(
+                () => Task.CompletedTask,
+                (next, action) => () => action.InvokeAsync(startupContext, next));
+
+        await moduleRunner.Start(moduleInstance, startupContext, pipeline);
+        moduleInstance.AddProperties(startupContext.GetProperties());
+        _lifetimeEvents.OnNext(new LifetimeEvent.Started(moduleInstance));
+
+        return moduleInstance;
+    }
+
+    public async Task StopModule(StopRequest request)
+    {
+        if (!_modules.TryGetValue(request.InstanceId, out var module))
+        {
+            return;
+        }
+
+        await StopModuleInternal(module);
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        _lifetimeEvents.Dispose();
```

Comment:
Move this to the end of the method? `StopModuleInternal` will still fire the lifetime events and maybe that's a good thing (we can't really make assumptions about when `DisposeAsync` is called)

Reply:
Let's do this as a follow-up